chore: update to marked version 9

robertIsaac opened the pull request:
### New features and enhancements

The property `markedExtensions` has been added to `MarkdownModuleConfig` allowing to use [marked extensions](https://marked.js.org/using_advanced#extensions) when configuring `MarkdownModule`.

```ts
MarkdownModule.forRoot({
  markedExtensions: [gfmHeadingId()],
}),
```

### ⚠ Breaking changes
- All options that were removed from marked has been deleted from this library too, see more at https://marked.js.org/using_advanced#options
- Some methods now return `Promise<string>` instead of `string`, because marked is doing so
- The `srcRelativeLink` input property is removed as the `baseUrl` option has been removed from marked, use https://www.npmjs.com/package/marked-base-url instead
- The `MarkdownPipe` now returns `Promise<string>` instead of `string` and will need to be combined with `async` pipe to work correctly

jfcere commented:
Hi @robertIsaac,

Thanks for the PR, I was expecting breaking changes, marked went from v5 to v9 in a few months only.

I'll review it and come back to you but at first glance it seems pretty descent!

UziTech commented on lib/src/markdown.service.ts:
```diff
@@ -423,4 +442,12 @@ export class MarkdownService {
+  private useExtensions() {
+    if (this.extensions) {
+      for (const ext of this.extensions) {
+        marked.use(ext);
```
You may want to think about using a [Marked instance](https://marked.js.org/using_advanced#instance) to prevent adding the same extensions multiple times. Some extensions will do strange things when added multiple times.

```js
import { Marked } from 'marked';

const marked = new Marked();

marked.use(...);
...
```

robertIsaac replied:
done

jfcere replied:
@UziTech instead of creating an instance of marked, we can just make sure that `marked.use(...)` is not called more than once right?

UziTech commented on lib/src/markdown.service.ts:
```diff
@@ -423,4 +442,12 @@ export class MarkdownService {
+
+  private useExtensions() {
+    if (this.extensions) {
+      for (const ext of this.extensions) {
```
the `marked.use` function can take multiple parameters that are extensions.

`marked.use(extension1, extension2)`

You can remove this for loop and use the spread operator
```js
marked.use(...this.extensions);
```
is the same as
```js
for (const ext of this.extensions) {
  marked.use(ext);
}
```

robertIsaac replied:
done and inlined it inside the `parseMarked` as you have suggested in the other discussion

UziTech commented on lib/src/marked-options.ts:
```diff
@@ -58,19 +33,4 @@ export class MarkedOptions {
    * Use smarter list behavior than the original markdown. May eventually be default with the old behavior moved into pedantic.
    */
   smartLists?: boolean;
```
`smartlists` was also removed a while ago, before the other options.

robertIsaac replied:
oh I thought it was something specific for `ngx-markdown` because I didn't see at all in marked docs
removed it as well now

robertIsaac commented:
hi @UziTech I have addressed your comments, please review again and resolve the conversation if my change address them correctly

robertIsaac commented:
hi @jfcere do you plan to merge this soon?

jfcere commented:
> hi @jfcere do you plan to merge this soon?

Because this is a break change, this cannot be merged in the master branch so I will have to create a release branch for the next major release.

Also, this the `README.md` file for documentation will have to be adjusted.

So, long story short, yes I will merge but this won't be available until Angular 17 is released which is due for next week according to their [release schedule](https://angular.io/guide/releases#release-schedule).

robertIsaac commented:
> > hi @jfcere do you plan to merge this soon?
> 
> Hi @robertIsaac,
> 
> First, let me say a big thank you for doing the migration, and also a special thanks to @UziTech for reviewing it!
> 
> Because this is a break change, this cannot be merged in the master branch so I will have to create a release branch for the next major release.
> 
> Also, the `README.md` file for documentation will have to be adjusted.
> 
> So, long story short, yes I will merge but this won't be available until Angular 17 is released which is due for next week according to their [release schedule](https://angular.io/guide/releases#release-schedule).

okay I will modify the `README.md` now
I understand that shouldn't be released now, but it should be merged (into master or release branch whatever you prefer, just tell me to change the target)
but I also want to make another PR to migrate everything to standalone but I'm afraid if I worked on it alongside this we might endup with conflicts

jfcere commented:
> okay I will modify the `README.md` now if you wish

Of course that would really be appreciated :)

robertIsaac commented:
another thing I want to do, to add prettier for the project if it's okay for you
I don't like that the every file has different format

robertIsaac commented:
> > okay I will modify the `README.md` now if you wish
> 
> Of course that would really be appreciated :)

done
removed all the options that were removed
added section for the new marked extension token

jfcere commented:
Hi @robertIsaac,

I've created a new branch for the next release and changed the target for your PR. As there were changes to update `mermaid.js` there is a small conflict in both `package.json` and `yarn.lock` file.

robertIsaac commented:
@jfcere done

jfcere commented:
@robertIsaac there are a few things I'd like to modify in your PR instead of pointing out everything and asking for you to do the changes. So I either do the changes directly in your PR or merge it and do another PR to adjust some stuff.

But seeing that you have another PR based on this one will surely bring conflicts to your other PR.. do you have a preference?

robertIsaac commented:
@jfcere  @michaelfaith
I made the suggestions here
also I found one unit test failed because using `Marked` class, did a workaround to fix it, I have no idea how to mock it directly using jasmine, if you have a better idea to mock it please share

robertIsaac commented:
uploading to coveralls is denied
I guess because the PR is made from my repo
do you have any suggestions what do ot @jfcere ?

jfcere commented:
> @jfcere @michaelfaith I made the suggestions here also I found one unit test failed because using `Marked` class, did a workaround to fix it, I have no idea how to mock it directly using jasmine, if you have a better idea to mock it please share

I just think that using a new instance of marked everytime is not necessary, only to call `marked.use()` once with the extensions, so this is a problem with the way the extensions are provided everytime in your implementation and I already have the fix... that I why I was saying I will either push modifications to your branch or merged it and do the changes afteward as your prefer otherwise it might end up in an endless chain of _"do this.. no that"_ etc

jfcere commented on demo/src/app/app.module.ts:
```diff
@@ -35,7 +35,6 @@ export function markedOptionsFactory(anchorService: AnchorService): MarkedOption
     ClipboardButtonModule,
     HttpClientModule,
     MarkdownModule.forRoot({
-      loader: HttpClient,
```
The `loader` property should not be removed.

This has been added so that people don't forget to import the `HttpClientModule` when they want to load markdown from external source. The same was as ngx-translate is doing it (see their example [here](https://github.com/ngx-translate/core#configuration).

robertIsaac replied:
it's different than ngx-translate
here is example
https://github.com/robertIsaac/angular-localization/blob/master/src/app/app.config.ts
I'm using a custom loader which you can find here https://github.com/robertIsaac/angular-localization/blob/master/src/app/core/utils/translate-browser.loader.ts
to change the path the files are in
and if someone forget it he will get a very clear message

that being said, if you wish to keep it, I will return it
and if that's the last thing to merge it you can return it and merge it directly instead of waiting for me

jfcere replied:
> it's different than ngx-translate here is example https://github.com/robertIsaac/angular-localization/blob/master/src/app/app.config.ts I'm using a custom loader which you can find here https://github.com/robertIsaac/angular-localization/blob/master/src/app/core/utils/translate-browser.loader.ts to change the path the files are in and if someone forget it he will get a very clear message
> 
> that being said, if you wish to keep it, I will return it and if that's the last thing to merge it you can return it and merge it directly instead of waiting for me

Because I had open issues about it previously and the fact that people could just go _"oh we are importing the HttpClientModule in our app but we aren't using it anywhere"_ I prefer to keep  it so that it makes it more obvious that it is used by the ngx-markdown library... the same way I prefer to have providers inside the MarkdownModule configuration instead of adding providers in the providers module property.

robertIsaac commented:
> > @jfcere @michaelfaith I made the suggestions here also I found one unit test failed because using `Marked` class, did a workaround to fix it, I have no idea how to mock it directly using jasmine, if you have a better idea to mock it please share
> 
> I just think that using a new instance of marked everytime is not necessary, only to call `marked.use()` once with the extensions, so this is a problem with the way the extensions are provided everytime in your implementation and I already have the fix... that is why I was saying I will either push modifications to your branch or merged it and do the changes afteward as your prefer otherwise it might end up in an endless chain of _"do this.. no that"_ etc

okay do your modifications in my branch I don't mind

robertIsaac commented:
> @robertIsaac there are a few things I'd like to modify in your PR instead of pointing out everything and asking for you to do the changes. So I either do the changes directly in your PR or merge it and do another PR to adjust some stuff.
> 
> But seeing that you have another PR based on this one, the changes will inevitably bring conflicts to your other PR... do you have a preference?

sorry I just saw this comment
do the changes so that we have one PR for the update
I will rebase the second PR once this merge is created

jfcere commented:
This PR will be merged into a temporary branch on ngx-markdown before merging into the next release branch to...

- Make adjustments as it is not impossible to override the renderer
- Merge the release branch with this PR changes
- Keep credits from the original PR author
